File: JavaScript/uri-1047.js
// URI - 1047: https://www.urionlinejudge.com.br/judge/pt/problems/view/1047

const input = require('fs').readFileSync('/dev/stdin', 'utf-8');
const values = input.split(' ').map(value => parseInt(value));

const startTime = { hour: values[0], minute: values[1] };
const endTime = { hour: values[2], minute: values[3] };
console.log(calculateGameDuration(startTime, endTime));

function calculateGameDuration(startTime, endTime) {
  const time = calculateDuration(startTime, endTime);
  return `O JOGO DUROU ${time.hours} HORA(S) E ${time.minutes} MINUTO(S)`;
}

function calculateDuration(startTime, endTime) {
  const startTimeInMinutes = convertTimeInMinutes(startTime);
  const endTimeInMinutes = convertTimeInMinutes(endTime);
  if (startTimeInMinutes === endTimeInMinutes)
    return { hours:24, minutes: 0 };
  else {
    const durationInMinutes = endTimeInMinutes - startTimeInMinutes;
    return convertMinutesInTime(durationInMinutes); 
  }  
}

function convertTimeInMinutes(time) {
  return (time.hour * 60) + time.minute;
}

function convertMinutesInTime(minutesToConvert) {
  let result;
  if (minutesToConvert > 0) {
    result = convertPositiveMinutesInTime(minutesToConvert)
  } else {
    result = convertNegativeMinutesInTime(minutesToConvert)
  }
  return result
}

function convertPositiveMinutesInTime(minutesToConvert) {
  const hours = minutesToConvert >= 60 
    ? parseInt(minutesToConvert / 60)
    : 0;
  const minutes = minutesToConvert - (hours * 60);
  return { hours, minutes }
}

function convertNegativeMinutesInTime(minutesToConvert) {
  let hours, minutes, hour;
  if (minutesToConvert <= -60) {
    hour = parseInt((Math.abs(minutesToConvert) / 60));
    hours = 23 - hour;
    minutes = 0
    Math.abs(minutesToConvert) === (hour * 60) 
      ? hours++ 
      : minutes = 60 - (Math.abs(minutesToConvert) - (hour * 60));  
  } else {
    hours = 23;
    minutes = minutesToConvert + 60;
  }
  return { hours, minutes }
}

File: Ruby/uri-1047.rb
# frozen_string_literal: true

# URI - 1047: https://www.urionlinejudge.com.br/judge/pt/problems/view/1047

def get_game_duration(time)
  game_duration = calculate_game_duration(time)
  if game_duration[:hours].zero? && game_duration[:minutes].zero?
    'O JOGO DUROU 24 HORA(S) E 0 MINUTO(S)'
  else
    "O JOGO DUROU #{game_duration[:hours]} HORA(S) E #{game_duration[:minutes]} MINUTO(S)"
  end
end

def calculate_game_duration(time)
  minutes = time[3] - time[1]
  hours = time[2] - time[0]
  hours = define_hours(hours, minutes)
  if minutes.negative?
    minutes += 60
    hours -= 1
  end
  { hours: hours, minutes: minutes }
end

def define_hours(hours, minutes)
  if hours.negative?
    24 - hours.abs
  elsif hours.zero? && minutes.negative?
    24
  else
    hours
  end
end

time = gets.split(' ').map(&:to_i)
puts get_game_duration(time)
